src/bosh-director/spec/unit/problem_resolver_spec.rb

```diff
@@ -31,96 +50,201 @@ def inactive_disk(id, deployment_id = nil)
+            resource_id: instance.id,
+            type: 'missing_vm',
+            state: 'open',
+          )
+          problem_resolutions[problem.id.to_s] = 'recreate_vm'
+        end
+      end
+
+      resolver = make_resolver(@deployment)
+      problem_handler = ProblemHandlers::MissingVM.new(1, nil)
+      allow(ProblemHandlers::Base).to receive(:create_from_model).and_return(problem_handler)
+
+      expect(problem_handler).to receive(:recreate_vm).exactly(problem_resolutions.size).times
+      expect(resolver.apply_resolutions(problem_resolutions)).to eq([problem_resolutions.size, nil])
+      expect(Models::DeploymentProblem.filter(state: 'open').count).to eq(0)
+    end
+
     describe '#apply_resolutions' do
       context 'when execution succeeds' do
+        let(:max_threads) { 10 }
+        let(:max_in_flight) { 5 }
+
         before do
-          allow(Bosh::Director::Config).to receive(:max_threads).and_return(5)
+          allow(Bosh::Director::Config).to receive(:max_threads).and_return(max_threads)
+          allow(parallel_update_config).to receive(:max_in_flight).and_return(max_in_flight)
         end
+
         context 'when parallel resurrection is turned on' do
-          it 'resolves the problems parallel' do
-            test_apply_resolutions
-            expect(ThreadPool).to have_received(:new).with(max_threads: 5)
-          end
-        end
+          context 'when only one problem exists per instance group' do
+            let(:n_igs_with_problems) { 1 }
+            let(:n_problems_in_ig) { 1 }
 
-        context 'when parallel resurrection is turned off' do
-          before do
-            allow(Bosh::Director::Config).to receive(:parallel_problem_resolution).and_return(false)
+            it 'does not create a threadpool for processing the problem' do
+              test_instance_apply_resolutions
+              expect(ThreadPool).not_to have_received(:new)
+            end
           end
-          it 'resolves the problems serial' do
-            test_apply_resolutions
-            expect(ThreadPool).not_to have_received(:new)
+
+          context 'when max_threads is one' do
+            let(:max_threads) { 1 }
+
+            it 'does not create a threadpool for processing the problem' do
+              test_instance_apply_resolutions
+              expect(ThreadPool).not_to have_received(:new)
+            end
           end
-        end
 
-        def test_apply_resolutions
-          disks = []
-          problems = []
+          context 'when max_in_flight is one' do
+            let(:max_in_flight) { 1 }
 
-          agent = double('agent')
-          expect(agent).to receive(:list_disk).and_return([])
-          expect(cloud).to receive(:detach_disk).exactly(1).times
+            it 'does not create a threadpool for processing the problem' do
+              test_instance_apply_resolutions
+              expect(ThreadPool).to have_received(:new).once.with(max_threads: n_igs_with_problems)
```

Comment:
I think this test description might be off, it seems like the code is asserting that ThreadPool is actually being created.

Also, shouldn't `ThreadPool` receive a new with a max_threads of what `max_threads` is set to? Why is the number of max_threads it's initialized with the number of instance groups with problems?

Reply:
Thanks for pointing this out! You are right, actually one ThreadPool should get created. But only the one which handles instance groups in parallel, therefore the number of `max_threads` is set to the number of instance_groups.